### PR TITLE
Refactor sepolicy.rules resolve

### PR DESCRIPTION
**Pull request by yujincheng08**
We resolve available partitions for sepolicy.rules when patching boot and bind mount the partition by magiskinit.

For older devices, the previous logic won't work because the part name is never readable.

### CODE REVIEW

**Comment by yujincheng08** [comment]
@HuskyDG no, it wont

File: native/src/init/mount.cpp [hunk 1]
```diff
@@ -117,99 +117,36 @@ static void switch_root(const string &path) {
...
-            return;
-        strcpy(p, MODULEROOT);
-        if (xaccess(path, F_OK) != 0) {
-            goto cache;
+    int major = 0;
+    int minor = 0;
+
+    parse_prop_file(".backup/.magisk", [&](auto key, auto value) -> bool {
+        if (key == "RULES_MAJOR") {
+            major = std::stoi(std::string(value));
```
Comment (canyie) [inline native/src/init/mount.cpp:125]:
```suggestion
            major = parse_int(value);
```

File: native/src/init/mount.cpp [hunk 2]
```diff
@@ -117,99 +117,36 @@ static void switch_root(const string &path) {
...
+        } else if (key == "RULES_MINOR") {
+            minor = std::stoi(std::string(value));
```
Comment (canyie) [inline native/src/init/mount.cpp:127]:
```suggestion
            minor = parse_int(value);
```

**Comment by vvb2060** [comment]
RULES_DEV=makedev(major, minor)

File: native/src/init/lib.rs [hunk 3]
```diff
@@ -1,11 +1,87 @@
...
+pub fn rules_device_config() -> i32 {
+    const UNKNOWN: i32 = 0;
+    const PERSIST: i32 = UNKNOWN + 1;
+    const METADATA: i32 = PERSIST + 1;
+    const CACHE: i32 = METADATA + 1;
+    const UNENCRYPTED: i32 = CACHE + 1;
+    const DATA: i32 = UNENCRYPTED + 1;
+    const OLD: i32 = DATA + 1;
+
+    let encrypted = env::var("ISENCRYPTED").map_or(false, |var| var == "true");
```
Comment (shymega) [inline native/src/init/lib.rs:27]: This PR was brought to my attention, and I just wanted to mention that this line isn't very Rust-like at all. There are better approaches to this code, and I thought I'd leave an amendment below:

```
let encrypted = env::var("ISENCRYPTED").unwrap_or(String::from("false"));
```

Presuming you want the default value to be "false". If not, you can change `String::from("false")` to `String::from("true")`.

You can then use traditional comparison operators on the `String`, or `.eq(T)`. 

I really like Magisk, so this is my way of just helping out with Rust side of things.

Comment (shymega) [reply native/src/init/lib.rs:27]: Having looked it again, I see something else. You're comparing it to a boolean. I didn't see that part clearly, it was nested. You could use `bool::from_str`, but I'm not convinced using `..map_or` is very Rusty. You're using `.map_or` to get a default value, right?

**Comment by vvb2060** [comment]
I don't think there's an advantage to parsing from native code, it's not performance sensitive code, and it's not going to be reused in native.
I'm all for rewriting shell code in native, but it should be a separate file, not moving methods to magiskinit and then calling them form shell, that's just weird.

File: native/src/init/lib.rs (hunk 3 above)
Comment (yujincheng08) [reply native/src/init/lib.rs:27]: @shymega 
1. `encrypted` is bool, but `env::var("ISENCRYPTED").unwrap_or(String::from("false"))` is `String`, and thus it's not correct.
2. if use `bool::from_str`, I have to `bool::from_str(&env::var("r").unwrap_or(String::from("false"))[..]).unwrap_or(false)`. I don't think this verbose way is better.

**Comment by Cyberfast2022** [comment]
Indeed, and I believe bootloops and boot failure won't end just there. That's true

File: native/src/init/lib.rs (hunk 3 above)
Comment (shymega) [reply native/src/init/lib.rs:27]: @yujincheng08 Yeah, I got the wrong impression. I think your solution, now I have re-read the PR, works fine. My apologies for the misunderstanding - your way is cleaner.

File: native/src/init/mount.cpp [hunk 4]
```diff
@@ -117,24 +117,12 @@ static void switch_root(const string &path) {
...
-            major = parse_int(value);
-        } else if (key == "RULES_MINOR") {
-            minor = parse_int(value);
-        }
-        return !major || !minor;
-    });
-
-    if (!major || !minor) return;
-    xmknod(BLOCKDIR "/rules", S_IFBLK | 0600, makedev(major, minor));
+    if (config->rules_dev == 0)
```
Comment (yujincheng08) [inline native/src/init/mount.cpp:120]: second stage wont load BootConfig and thus this is always false

**Review by topjohnwu — APPROVED** [review]

**Pull request merged**